stagelib/replace.py

```diff
@@ -0,0 +1,106 @@
+GLOBAL_PATTERN = re.compile('STAGE_REPLACE_WITH_[A-Z\_]+')
+
+logger = logging.getLogger(__name__)
+
+
+# Read text from the file
+def _get_file_text(file_path):
+    with open(file_path, 'r') as txtfile:
+        return txtfile.read()
+
+
+# Given a variable file, read the variable file, and replace it's corresponding
+# variable in the text with the contents of the file without trailing space
+def _replace_file_in_text(variable_file_path, text, file_path):
+    variable = os.path.basename(variable_file_path)
+    with open(variable_file_path) as variable_file:
+        variable_text = variable_file.read().rstrip()
+        if "\n" in variable_text:
+            # Output 'user-defined script' if string is multiline
+            logging.info('\t{:<80}  :: {:<38} <- {}'.format(file_path, variable, '[user-defined script]'))
+        else:
+            logging.info('\t{:<80}  :: {:<38} <- {}'.format(file_path, variable, variable_text))
+        return text.replace(variable, variable_text)
+
+
+# Move existing file to <file>.bak. Save text into file.
+def _save_with_backup(file_path, text):
+    os.rename(file_path, '{}.bak'.format(file_path))
+    with open(file_path, 'w') as new_file:
+        new_file.write(text)
+
+
+# Perform
+def _file_replace(template_text, file_path, variable_file_dir):
+    variable_matches = re.findall(VARIABLE_FILE_PATTERN, template_text)
+    if not variable_matches:
+        logging.debug('\t{:<80}     [No __VARIABLES__ to replace]'.format(file_path))
+        return template_text
+    # Allow variable files to have their own variables in them
+    # by reprocessing until there are no more matches
+    while variable_matches:
+        for var in variable_matches:
+            template_text = _replace_file_in_text(os.path.join(variable_file_dir, var),
+                                                  template_text, file_path)
+        variable_matches = re.findall(VARIABLE_FILE_PATTERN, template_text)
+    return template_text
+
+
+def _replace_global_in_text(global_var_match, text):
+    global_var_name = global_var_match[len('STAGE_REPLACE_WITH_'):]  # strip STAGE_REPLACE_WITH_
+    global_value = globals()[global_var_name]
+    return text.replace(global_var_match, global_value)
+
+
+def _global_replace(template_text, file_path):
+    variable_matches = re.findall(GLOBAL_PATTERN, template_text)
+    if not variable_matches:
+        # logging.debug('\t{:<80}  -> {}'.format('No GLOBAL_VARs', file_path))
+        logging.debug('\t{:<80}     [No STAGE_REPLACE_WITH_VARs to replace]'.format(file_path))
+        return template_text
+    for var in variable_matches:
+        var_name = var[len('STAGE_REPLACE_WITH_'):]  # strip STAGE_REPLACE_WITH_
+        logging.info('\t{:<80}  :: {:<38} <- {}'.format(file_path, var_name, globals()[var_name]))
+        template_text = _replace_global_in_text(var, template_text)
+    return template_text
+
+
+def do_variable_replace(replace_root_dir):
```

Comment:
this looks like a template engine. Why this PR is not proposing to use a well-established templating engine to do this work instead of adding untested (and undocumented) code?

Reply:
@alfredodeza What do you have in mind ?
We have some requirements like keeping a file-based mecanism and we don't have any yaml/xml stuff.

Reply:
I had a discussion about this with @ErwanAliasr1. I'm open to suggestions, but I don't know of any templating engines that meet our needs of being able to have inputs defined by files that can be overridden by more specific files. In our discussion, we reasoned that we are reinventing a wheel to some perspective, but it is a lightweight engine that is very straightforward with no need to hack around forcing our override system to support inputs to an outside template engine. We can simply put text in a file with no extra boilerplate, and this will be easier for newcomers.

Reply:
@ErwanAliasr1 we've advocated in the past for a simplification of the current multi-distro multi-symlink process. The idea of trying to simplify is correct. What we've asked many times before is to *not* support a matrix of distributions because there is almost no value and it generates a tremendous burden in maintenance. The complexity in this PR to make it simpler to keep the matrix around should be another strong point in favor of just not supporting multi-distro generation.

This also does not make it any simpler for any downstream process.

Reply:
@BlaineEXE this will absolutely not be easier for newcomers, as it is highly undocumented, tested with a non-standard bash script, and it tries to solve a very complicated arrangement of multi-distro building for the project. This is not straightforward at all

Reply:
Documentation can indeed be improved. Feel free to help with that.

As for not supporting multiple distros, I'd like to see what @leseb, @liewegas, and @dmick think. I agree that not supporting multiple distros will make it easier to support. It will also make it harder to support users who would like to build for other distros.

Reply:
@alfredodeza that project does improve the way that downstream works as it put a real split between what is common and what is different from every distro. A downstream should only work in its distro directory.
And I'm sorry but I don't understand your comment here. We have to support several ceph releases, on several distro and several versions of those. That makes a matrix in anyway.  This solution clearly expose what is common and what is different between flavors. There is not a single link avoiding then colateral damage when you edit a file. That also kills code duplication we had.
So yes, that PR needs to be more documented, and we'll do but it does answer of maintainability purpose.  I can hear that's not so obvious for now, we'll work on clarifying that.

Reply:
@alfredodeza If you're interested in learning more about multi-distro support, you should read https://github.com/ceph/ceph-container/issues/247. It gives the whole perspective about distro management and why it's essential/crucial.

We support multi-distro for builds **only**, we are not testing the code on all the distributions.

Reply:
@leseb yes, I've read that isssue before. What is a supported distro? Something where code exists to "create" a version of it? There aren't tests nor any sort of validation. What is the benefit of the "community" for something that is not tested? If it is essential/crucial, why aren't we expanding the matrix to really test every single permutation that this project attempts to support? 

What does it mean to "support multi-distro for builds only" ? Again, are you verifying these builds work? where are the tests for these? What is the use of a build for a project that we don't know if it runs?

This PR changes over six hundred files trying to improve the build situation. This is not a solution, it is a symptom of the larger problem.

The Ceph project for the Luminous release only builds (and tests!) Ubuntu Xenial, and CentOS7.

Reply:
My $0.02: don't let perfect be the enemy of good (or even just better).

We need multi-distro support.  18.04 will be added shortly, and downstream vendors will be adding in their own distros as well.

We need multi-version support of some form.  We'll be building containers for luminous, soon mimic, and probably even jewel for some time.  That could be maintained in a parallel for each release but there appears to be value in this approach that automates some of that maintenance. 

Whether all distro combinations are tested seems like an orthogonal issue.  There is value in having an untested build (even if it is not as much value as a tested one), and having a tested distro target that is close to what you need (e.g. tested ubuntu vs untested debian, tested centos vs untested rhel).

Reply:
thx @liewegas for sharing this vison. We do need all downstreams to be part of the effort. We are here, Suse does. That's a good start. I cannot see how downstreams could accept shipping containers without having their own stuff inside. This PR is about ease this process by having a clear segregation between upsteam & downstream code.

Reply:
on the template engine question: it seems odd to me that we're reinventing variable substitution here too.  The precedence definition/mechanism is outside the scope of the template engine.

Reply:
If you have a template engine in mind, please suggest one. I am somewhat familiar with Jinja 2, but if we use it, templates must be filled by yaml files. Doing yaml file overriding of specific values without deleting different, existing variables seems non-trivial to me. Furthermore, my proposed implementation supports nested variables. I wasn't sure what Jinja's support of that would be. It seems easier to write a few python lines just to read a file for the replacement. And it was an extra 3 lines of code to support nested variables. It was the fastest path to success I found, and there are quite extensive integration tests for all of the features of this script to verify that it works as expected.

Reply:
I don't think jinja2 even knows what data file input format you might use, much less yaml; you call render() with a set of kv pairs, and it's up to you where those come from and how overriding is done; I would assume most templating libs do the same.  And a 'merge' operation is pretty standard for multiple configuration sources; you just define order of precedence for duplicated keys.

I don't have any alternate design fleshed out in my head, it was just surprising to see basic variable replacement in text files *not* using an existing solution.  I don't yet really understand the nestedness